networkit/cpp/embedding/LearnEmbedding.cpp

```diff
@@ -0,0 +1,256 @@
+constexpr double startAlpha = 0.025;
+constexpr int refreshAlphaCount = 10000;
+constexpr double minAlpha = startAlpha * 0.0001;
+constexpr double nsPwr = 0.75;
+/*
+The exponent used to shape the negative sampling distribution.
+A value of 1.0 samples exactly in proportion to the frequencies,
+0.0 samples all words equally, while a negative value samples
+low-frequency words more than high-frequency words.
+The popular default value of 0.75 was chosen by the original Word2Vec paper.
+*/
+
+using Vocab = std::vector<count>;
+
+Vocab learnVocab(const AllWalks &allWalks, count nn) {
+    Vocab vocab(nn);
+    for (auto &walk : allWalks) { // for every walk starting from this node
+        for (auto &node : walk) { // for every node in this walk
+            ++vocab[node];
+        }
+    }
+    return vocab;
+}
+
+// Precompute unigram table using alias sampling method
+AliasSampler vocabSampler(const Vocab &vocab) {
+    double trainWordsPow = 0;
+    auto sz = vocab.size();
+    std::vector<float> probV(sz); // NOT ... {sz}
+    for (count i = 0; i < sz; ++i) {
+        double e = std::pow(vocab[i], nsPwr);
+        probV[i] = e;
+        trainWordsPow += e;
+    }
+    std::transform(probV.begin(), probV.end(), probV.begin(),
+                   [&trainWordsPow](float f) { return f / trainWordsPow; });
+
+    AliasSampler as(sz);
+    as.unigram(probV);
+    return as;
+}
+
+// Initialize negative embeddings
+Embeddings initNegEmb(const Vocab &vocab, int dimensions) {
+    return Embeddings{vocab.size(), std::vector<float>(dimensions)};
+}
+
+// Initialize positive embeddings
+Embeddings initPosEmb(const Vocab &vocab, int dimensions) {
+    Embeddings emb{vocab.size(), std::vector<float>(dimensions)};
+#pragma omp parallel for schedule(dynamic)
+    for (omp_index i = 0; i < static_cast<omp_index>(emb.size()); ++i) {
+        std::generate(emb[i].begin(), emb[i].end(),
+                      [&]() { return (Aux::Random::real() - 0.5) / dimensions; });
+    }
+    return emb;
+}
+
+void trainModel(const ModelData &model, count walkNr) {
+    AllWalks &allWalks = model.allWalks;
+    count &wordCntAll = model.wordCntAll;
+    AliasSampler &as = model.as;
+    count allWords = model.allWords;
+    count iterations = model.iterations;
+    count dimensions = model.dimensions;
+    count winSize = model.winSize;
+    double alpha = model.alpha;
+    Embeddings &synPos = model.synPos;
+    Embeddings &synNeg = model.synNeg;
+
+    using Walk = BiasedRandomWalk::Walk;
+
+    Walk &thisWalk = allWalks[walkNr];
+
+    for (index wordI = 0; wordI < thisWalk.size(); ++wordI) {
+        if (wordCntAll % refreshAlphaCount == 0) {
+#ifndef NETWORKIT_OMP2
+#pragma omp atomic write
+#endif // NETWORKIT_OMP2
+            alpha = startAlpha * (1 - wordCntAll / ((double)(iterations * allWords) + 1.0));
+            if (alpha < minAlpha) {
+#ifndef NETWORKIT_OMP2
+#pragma omp atomic write
+#endif // NETWORKIT_OMP2
+                alpha = minAlpha;
```

Comment:
If you change this to a reference, additional care needs to be taken to avoid that other threads read a value of `alpha` below `minAlpha`. For this, the check for `minAlpha` needs to be before assigning the new value to `alpha` (i.e., first store the value in a local temporary, compare, and then assign either the temporary or `minAlpha`). I am not sure though if this needs to be a model parameter at all as it seems to be recomputed from scratch anyway and is not used afterwards.

Reply:
Yes, it should be a reference. The original code from snap does no protection against parallel accesses at all - so did I.
So I have the choice to take all these effects as additional randomness OR implement a version, which is not the 'same' as the original!? In discussion with Alexander van der Grinten we concluded, that it should be sufficient to prevent non-atomic writes.